sound/soc/sof/sof-pci-dev.c

```diff
@@ -222,20 +222,6 @@ static int sof_pci_probe(struct pci_dev *pci,
 #else
 	/* find machine */
 	mach = snd_soc_acpi_find_machine(desc->machines);
```

Comment:
I don't get the flow here. For HDaudio this will fail, so how do you not end-up in the error case.
Even if you set the mach variable earlier, it'll be overridden here.
And you are doing this before doing the DSP probe (the platform driver is registered later)

Does this even work?

Reply:
For HDaudio, nocodec is setup here.  in sof_probe, if hd codec is detected , HDaudio machine driver would be applied. 

platform driver is registered after HDaudio machine driver is applied.

I test it on WHL, it works.

Reply:
@RanderWang 
I don't get what you are saying and 'it works' isn't a good explanation, sorry I don't know what 'it' means. 
The ask is that we can have a single kernel with HDaudio compiled in, and the behavior is as follows
a. try to find an ACPI ID and create the relevant machine driver
b. if this fails, check if there are HDAudio codecs. if HDaudio + iDISP are found, use the generic HDaudio machine driver.
c1. if nothing is detected fall back to nocodec. 
c2. If only iDISP is found, either fallback to a nocodec+iDISP configuration or to a simple iDISP configuration if nocodec is not selected.

Is this clearer?